polaris/evaluate/_metric.py

```diff
@@ -84,55 +164,241 @@ class Metric(Enum):
         assert metric.score(y_true=first, y_pred=second) == metric(y_true=first, y_pred=second)
         ```
         """
-        if y_pred is None and y_prob is None:
-            raise ValueError("Neither `y_pred` nor `y_prob` is specified.")
+        return self.score(y_true, y_pred, y_prob)
 
-        if self.y_type == "y_pred":
-            if y_pred is None:
-                raise ValueError(f"{self} requires `y_pred` input. ")
-            pred = y_pred
-        else:
-            if y_prob is None:
-                raise ValueError(f"{self} requires `y_prob` input. ")
-            pred = y_prob
+    # For any given benchmark, all of its metrics need to have a unique name.
+    # Within the scope of a given benchmark metric is thus uniquely identified by its name.
 
-        kwargs = {"y_true": y_true, self.y_type: pred}
-        return self.fn(**kwargs, **self.value.kwargs)
+    def __hash__(self) -> int:
+        return hash(self.name)
 
-    def __call__(
-        self, y_true: np.ndarray, y_pred: np.ndarray | None = None, y_prob: np.ndarray | None = None
+    def __eq__(self, other) -> bool:
+        match other:
+            case Metric():
+                return other.name == self.name
+            case _:
+                return False
+
+
+class GroupedMetricConfig(BaseModel):
+    """
+    The configuration for a GroupedMetric
+
+    Attributes:
+        group_by: The column to group by.
+        on_error: How to handle errors when computing the metric.
+        default: The default value to use when an error occurs.
+        aggregation: The aggregation method to use when combining the metric scores.
+    """
+
+    group_by: str
+    on_error: Literal["ignore", "raise", "default"] = "raise"
+    default: float | None = None
+    aggregation: Literal["mean", "median"] = "mean"
+
+
+class GroupedMetric(Metric):
+    """
+    A GroupedMetric is a Metric that computes the metric separately for each group in a dataset.
+    A final score is then computed by aggregating the scores for each group.
+
+    Warning: Memory usage
+        The current implementation of GroupedMetric relies on a Pandas DataFrame and loads
+        the entire test set (including inputs) into memory. It is therefore not suited for large benchmarks.
+
+    Attributes:
+        kind: The kind of metric, uses to discriminate different types of metrics in a discriminated union.
+        config: The configuration for the grouped metric.
+    """
+
+    kind: Literal["grouped"] = "grouped"
+    config: GroupedMetricConfig
+
+    @model_validator(mode="after")
+    def check_default_required(self) -> Self:
+        """Check that a default value is provided when on_error is 'default'."""
+        handling = self.config.on_error
+        default = self.config.default
+        if handling == "default" and default is None:
+            raise ValueError("Default value must be provided when on_error is 'default'")
+        return self
+
+    def score(
```

Comment:
**Suggestion**
Move this into either the parent class or have it together with the config in one class.

- If into the parent class, have private `score` methods for each kind (default, grouped) of metrics.
- If into together with the config, have the parent class call it when config is present, with the default `solve` of no config.